File: 251004/Snopko/src/main/java/com/example/restservice/service/UserService.java
package com.example.restservice.service;

import com.example.restservice.dto.request.UserRequestTo;
import com.example.restservice.dto.response.UserResponseTo;
import com.example.restservice.model.User;
import com.example.restservice.repository.UserRepository;

import lombok.AllArgsConstructor;
import org.modelmapper.ModelMapper;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;
import java.util.NoSuchElementException;
import java.util.stream.Collectors;

@AllArgsConstructor
@Service
public class UserService {
    private final UserRepository userRepository;
    private final ModelMapper modelMapper;

    public List<UserResponseTo> getAll() {
        return userRepository.findAll().stream()
                .map(this::toResponse)
                .collect(Collectors.toList());
    }

    public UserResponseTo getById(Long id) {
        return userRepository.findById(id)
                .map(this::toResponse)
                .orElseThrow(() -> new NoSuchElementException("User not found"));
    }

    public UserResponseTo create(UserRequestTo request) {
        User user = modelMapper.map(request, User.class);
        User saved = userRepository.save(user);
        return toResponse(saved);
    }

    public UserResponseTo update(UserRequestTo request) {
        if (request.getId() == null) {
            throw new NoSuchElementException("User not found");
        }

        User existing = userRepository.findById(request.getId())
                .orElseThrow(() -> new NoSuchElementException("User not found"));

        modelMapper.map(request, existing);
        User updated = userRepository.save(existing);
        return toResponse(updated);
    }

    public void delete(Long id) {
        if (userRepository.findById(id).isEmpty()) {
            throw new NoSuchElementException("User not found");
        }
        userRepository.deleteById(id);
    }

    private UserResponseTo toResponse(User user) {
        return modelMapper.map(user, UserResponseTo.class);
    }
}

File: 251003/Kukhatskavolets/microblog/src/main/kotlin/by/kukhatskavolets/services/UserService.kt
package by.kukhatskavolets.services

import by.kukhatskavolets.dto.requests.UserRequestTo
import by.kukhatskavolets.dto.responses.UserResponseTo
import by.kukhatskavolets.mappers.toEntity
import by.kukhatskavolets.mappers.toResponse
import by.kukhatskavolets.repositories.TweetRepository
import by.kukhatskavolets.repositories.UserRepository
import org.springframework.stereotype.Service

@Service
class UserService(
    private val userRepository: UserRepository,
    private val tweetRepository: TweetRepository,
) {
    fun createUser(userRequestTo: UserRequestTo): UserResponseTo {
        val user = userRequestTo.toEntity()
        val savedUser = userRepository.save(user)
        return savedUser.toResponse()
    }

    fun getUserById(id: Long): UserResponseTo {
        val user = userRepository.findById(id).orElseThrow { NoSuchElementException() }
        return user.toResponse()
    }

    fun getAllUsers(): List<UserResponseTo> =
        userRepository.findAll().map { it.toResponse() }

    fun updateUser(id: Long, userRequestTo: UserRequestTo): UserResponseTo {
        val updatedUser = userRequestTo.toEntity().apply { this.id = id }
        return userRepository.save(updatedUser).toResponse()
    }

    fun deleteUser(id: Long) {
        userRepository.findById(id).orElseThrow { NoSuchElementException() }
        userRepository.deleteById(id)
    }

    fun getUserByTweetId(tweetId: Long): UserResponseTo? {
        val tweet = tweetRepository.findById(tweetId).orElseThrow { NoSuchElementException() }
        return tweet.user.toResponse()
    }
}